sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update.hpp

```diff
@@ -0,0 +1,364 @@
+  return out;
+}
+
+const char *to_string(atomic_op op) {
+  switch (op) {
+  case atomic_op::add:
+    return "add";
+  case atomic_op::sub:
+    return "sub";
+  case atomic_op::inc:
+    return "inc";
+  case atomic_op::dec:
+    return "dec";
+  case atomic_op::umin:
+    return "umin";
+  case atomic_op::umax:
+    return "umax";
+  case atomic_op::xchg:
+    return "xchg";
+  case atomic_op::cmpxchg:
+    return "cmpxchg";
+  case atomic_op::bit_and:
+    return "bit_and";
+  case atomic_op::bit_or:
+    return "bit_or";
+  case atomic_op::bit_xor:
+    return "bit_xor";
+  case atomic_op::smin:
+    return "smin";
+  case atomic_op::smax:
+    return "smax";
+  case atomic_op::fmax:
+    return "fmax";
+  case atomic_op::fmin:
+    return "fmin";
+  case atomic_op::fadd:
+    return "fadd";
+  case atomic_op::fsub:
+    return "fsub";
+  case atomic_op::fcmpxchg:
+    return "fcmpxchg";
+  case atomic_op::load:
+    return "load";
+  case atomic_op::store:
+    return "store";
+  case atomic_op::predec:
+    return "predec";
+  }
+  return "<unknown>";
+}
+
+template <int N> inline bool any(simd_mask<N> m, simd_mask<N> ignore_mask) {
+  simd_mask<N> m1 = 0;
+  m.merge(m1, ignore_mask);
+  return m.any();
+}
+
+// ----------------- The main test function
+template <class T, int N, template <class, int> class ImplF, bool UseProperties>
+bool test(queue q, const Config &cfg) {
+  constexpr auto op = ImplF<T, N>::atomic_op;
+  using CurAtomicOpT = decltype(op);
+  constexpr int n_args = ImplF<T, N>::n_args;
+
+  std::cout << "USM Testing " << "op=" << to_string(op) << "n_args=" << n_args
+            << " T=" << esimd_test::type_name<T>() << " N=" << N << "\n\t"
+            << cfg << "...";
+
+  size_t size = cfg.start_ind + (N - 1) * cfg.stride + 1;
+  T *arr = malloc_shared<T>(size, q);
+  int n_threads = cfg.threads_per_group * cfg.n_groups;
+
+  for (int i = 0; i < size; ++i) {
+    arr[i] = ImplF<T, N>::init(i, cfg);
+  }
+
+  range<1> glob_rng(n_threads);
+  range<1> loc_rng(cfg.threads_per_group);
+  nd_range<1> rng(glob_rng, loc_rng);
+
+  properties props{cache_hint_L1<cache_hint::uncached>,
+                   cache_hint_L2<cache_hint::write_back>};
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for(rng, [=](id<1> ii) SYCL_ESIMD_KERNEL {
+        int i = ii;
+        simd<Toffset, N> offsets(cfg.start_ind * sizeof(T),
+                                 cfg.stride * sizeof(T));
+        simd_mask<N> m = 1;
+        if (cfg.masked_lane < N)
+          m[cfg.masked_lane] = 0;
+        // barrier to achieve better contention:
+        // Intra-work group barrier.
+        barrier();
+
+        // the atomic operation itself applied in a loop:
+        for (int cnt = 0; cnt < cfg.repeat; ++cnt) {
+          if constexpr (n_args == 0) {
+            if constexpr (UseProperties)
+              atomic_update<op>(arr, offsets, m, props);
+            else
+              atomic_update<op>(arr, offsets, m);
```

Comment:
There should be 4 variants of calls: 2 more are without mask 'm'.

Reply:
Done in https://github.com/intel/llvm/pull/11833/commits/ffe19702c45355afd8b7ab17d2170cd783e9f670.